_posts/2016/2016-11-21-multitenancy-and-elasticsearch.md

```diff
@@ -0,0 +1,144 @@
+As an example, for 10,000 tenants, 10 types and 10 supported languages, you will need 1 million indices.
+
+### Shared Index
+
+To keep the amount of indices under control, you can share the same index for all tenants.
+This would reduce the number of indices to 100 for the example mentioned above.
+
+Elasticsearch provides some features that help dealing with multitenancy in a single index.
+
+#### Bool / Filtered query
+
+An obvious way to implement this is by wrapping queries into a [bool query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-bool-query.html) and filter by tenant.
+This is very similar to what you might be used to with other databases and corresponding frameworks.
+
+{% highlight javascript %}
+POST /products/_search
+{
+  "query" : {
+    "bool" : {
+      "filter" : {
+        "term" : { "tenantId" : "12345" }
+      },
+      // ...
+    }
+  }
+}
+{% endhighlight %}
+
+This is a possible solution, but you would probably like to have some kind of framework support for that, otherwise you forget to wrap a query somewhere and end up exposing data across tenants.
+
+#### Index alias
+
+Another nifty feature you can use are [filtered aliases](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-aliases.html#filtered).
+
+Filtered aliases basically enable you to do the same thing as with filtered queries by introducing an alias for every tenant.
+This way you can deal with tenancy in a more transparent way.
+You don't need to wrap every query yourself, you let Elasticsearch do it for you.
+
+{% highlight javascript %}
+POST /_aliases
+{
+  "actions" : [
+    {
+      "add" : {
+         "index" : "products",
+         "alias" : "products_12345",
+         "filter" : { "term" : { "tenantId" : "12345" } }
+      }
+    }
+  ]
+}
+{% endhighlight %}
+
+{% highlight javascript %}
+POST /products_12345/_search
+{
+  "query" : {
+    // ...
+  }
+}
+{% endhighlight %}
+
+It is also more secure when it comes to data separation, because you will need to specify the index/alias you want to query.
+Forgetting about the tenant will lead to the index not being found instead of retrieving the results for all tenants.
+
+By delegating this responsibility to Elasticsearch, you can get rid of some complexity in your code, but you need to take care of creating and deleting these aliases.
+In our architecture, a service emits events for shop activation/deactivation and we listen for those to create/delete aliases automatically.
+
+Aliases are pretty lightweight, but in extreme cases this might still lead to performance issues.
+If you create too many aliases, you increase the size of your cluster state and might introduce a new bottleneck.
+The cluster state is transferred between nodes on every update and includes index, alias and mapping information.
+
+Another issue with using a shared index in general is that there is no complete data separation for scoring.
```

Comment:
a comma after "in general"? -> "Another issue with using a shared index in general, is ..."